src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py

```diff
@@ -0,0 +1,519 @@
+            with. Pass `"self"` to use self distillation with a teacher
+            from a copy of the given module at the start epoch.
+            Pass `"disable"` to disable this modifier.
+            Defaults to `"disable"`
+        :param kwargs: Optional kwargs to support specific arguments
+            for individual modifiers.
+        """
+        super().initialize(module, epoch, loggers, distillation_teacher, **kwargs)
+
+        if not self._distillation_enabled:
+            return
+
+        if distillation_teacher == "self":
+            self._teacher = deepcopy(module)
+
+        self._reset_cache()
+
+        cached_student_layers: Dict[str, torch.nn.Module] = {}
+        if self._student_layer_names is None:
+            _update_layers_by_type(module, cached_student_layers)
+            self._student_layer_names = list(cached_student_layers.keys())
+        else:
+            _update_layers_by_name(
+                module, self._student_layer_names, cached_student_layers
+            )
+        _LOGGER.info("Distilling student layers: %s", self._student_layer_names)
+
+        cached_teacher_layers: Dict[str, torch.nn.Module] = {}
+        if self._teacher_layer_names is None:
+            _update_layers_by_type(self._teacher, cached_teacher_layers)
+            self._teacher_layer_names = list(cached_teacher_layers.keys())
+        else:
+            _update_layers_by_name(
+                self._teacher, self._teacher_layer_names, cached_teacher_layers
+            )
+        _LOGGER.info("Distilling teacher layers: %s", self._teacher_layer_names)
+
+        if len(self._teacher_layer_names) != len(self._student_layer_names):
+            raise ValueError(
+                "Found different numbers of teacher and student layers to distill. "
+                "Set teacher_layer_names and student_layer_names explicitly."
+            )
+
+        self._student_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_student_output, self._student_output_shapes
+                )
+            )
+            for name, layer in cached_student_layers.items()
+        ]
+
+        self._teacher_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_teacher_output, self._teacher_output_shapes
+                )
+            )
+            for name, layer in cached_teacher_layers.items()
+        ]
+
+    def update(
+        self,
+        module: Module,
+        optimizer: torch.optim.Optimizer,
+        epoch: float,
+        steps_per_epoch: int,
+    ):
+        super().update(module, optimizer, epoch, steps_per_epoch)
+        if epoch >= self.end_epoch:
+            for handle in self._student_handles:
+                handle.remove()
+            for handle in self._teacher_handles:
+                handle.remove()
+            self._reset_cache()
```

Comment:
If I understand it correctly, _reset_cache is clearing the projection parameters. We want these parameters to be saved w/ the model and to be able to reload them